src/devices/src/virtio/mmio.rs

```diff
@@ -204,18 +195,23 @@ impl MmioTransport {
             }
             _ if status == 0 => {
                 if self.device_activated {
-                    let mut device = self.device.lock().expect("Poisoned device lock");
-                    match device.reset() {
+                    let mut device_activated = true;
+                    let mut device_status = self.device_status;
+
+                    match self.locked_device().reset() {
                         Some((_interrupt_evt, mut _queue_evts)) => {
-                            self.device_activated = false;
+                            device_activated = false;
                         }
                         // Backend device driver doesn't support reset,
                         // just mark the device as FAILED.
                         None => {
-                            self.device_status |= FAILED;
+                            device_status |= FAILED;
```

Comment:
The `FAILED` flag is lost here. Suggest changing it back to `self.device_status |= FAILED`.

Reply:
Sounds good.

Just as a note: I think the double borrow could be avoided:
```rust
let reset_result = self.locked_device().reset();
match reset_result {
    // free to use `self` here.
}
```